### ISSUE TITLE
Add support for DOMDocument::registerNodeClass() polymorphism?

### ISSUE THREAD

**Issue by Girgias** [vimeo/psalm#9343]
This is definitely in the realm of strange, and maybe impossible, feature requests. So feel free to close if this sounds unreasonable.

I am currently working on an XML [DocBook renderer](https://gitlab.com/Girgias/docbook-renderer), and I'm using the ``DOMDocument::registerNodeClass()`` method to overload the classes that are returned by the DOM PHP extension: https://gitlab.com/Girgias/docbook-renderer/-/blob/master/src/DOMRenderingDocument.php

Psalm is complaining, and for good reasons, that the ``render()`` method does not exist on ``DOMNode``.

In the same vein, I get errors like:
```
ERROR: ArgumentTypeCoercion - src/Tags/QandARenderer.php:56:63 - Argument 1 of Girgias\DocbookRender\Tags\SimpleReplacementTagRender::__construct expects Girgias\DocbookRender\DOMRenderingElement, but parent type DOMNode provided (see https://psalm.dev/193)
            $entryRenderer .= (new SimpleReplacementTagRender($subNode, 'dd', classes: ['answer']))->render($state);
```

Which make sense, however, I know those are wrong because the ``$subNode`` comes from ``$this->element->childNodes`` which will return a collection of ``DOMRenderingElement|DOMRenderingText``.

Moreover, I don't hit into any TypeErrors at run-time.

Is there a way, other than adding a bunch of doc comments, to annotate the types returned by my overloaded DOMNode objects? Or is this just an issue with the design I'm currently using.

**Comment by Girgias** [comment]
https://psalm.dev/r/137576313f

**Comment by weirdan** [comment]
It would mean DOMDocument needs to have type parameters for all those node classes. While that's not impossible in theory, it would be quite impractical for all those who *don't* use `registerNodeClass()`, as Psalm does not have a concept of defaults for type parameters. Every time DOMDocument is instantiated, people would have to use var docblock like

```php
/** @var DomDocument<DOMNode, DOMAttr, DOMElement, DOMComment, DOMProcessingInstruction, DOMText> */
$doc = new DOMDocument(.....);
```
(and I probably missed a few). 

Perhaps this could be done in custom code (e.g. a plugin) that would register a non-standard stub with all those type parameters, set those parameters to default values when `DOMDocument` is instantiated and updated them when `registerNodeClass()` is called.
